pg_search/src/postgres/storage/avl.rs

```diff
@@ -0,0 +1,1230 @@
+                        self.arena[i].key = succ_key;
+                        self.arena[i].val = succ_val;
+                        let temp_tag = self.arena[i].tag; // Save original tag of slot i
+                        self.arena[i].tag = succ_tag; // Move successor's tag to slot i
+                        self.arena[succ].tag = temp_tag; // Move original tag to successor's slot
+
+                        // Delete successor, but do NOT overwrite removed_val
+                        let mut ignored_val: Option<V> = None;
+                        let mut ignored_flag = false;
+                        let right = self.delete_rec(
+                            self.arena[i].r(),
+                            &succ_key,
+                            &mut ignored_val,
+                            &mut ignored_flag,
+                        );
+                        self.arena[i].set_r(right);
+
+                        Some(self.rebalance(i))
+                    }
+                }
+            }
+        }
+    }
+}
+
+const MAX_HEIGHT: usize = 128;
+
+pub struct AvlIter<'a, K: Copy, V: Copy, T: Copy> {
+    arena: &'a [Slot<K, V, T>],
+    stack: [usize; MAX_HEIGHT],
+    sp: usize,
+    cursor: Option<usize>,
+}
+
+impl<'a, K: Ord + Copy, V: Copy, T: Copy> AvlTreeMapView<'a, K, V, T> {
+    /// In-order iterator over (&copy) (K, V) pairs.
+    #[inline]
+    pub fn iter(&'a self) -> AvlIter<'a, K, V, T> {
+        AvlIter {
+            arena: self.arena,
+            stack: [0; MAX_HEIGHT],
+            sp: 0,
+            cursor: self.header.root,
+        }
+    }
+}
+
+impl<'a, K: Copy, V: Copy, T: Copy> AvlIter<'a, K, V, T> {
+    #[inline]
+    fn push_left_chain(&mut self) {
+        // Descend left from `cursor`, pushing nodes onto the fixed stack.
+        while let Some(i) = self.cursor {
+            debug_assert!(self.sp < MAX_HEIGHT, "AVL height exceeded MAX_HEIGHT");
+            self.stack[self.sp] = i;
+            self.sp += 1;
+            self.cursor = self.arena[i].l();
+        }
+    }
+}
+
+impl<'a, K: Ord + Copy, V: Copy, T: Copy> Iterator for AvlIter<'a, K, V, T> {
+    type Item = (K, V);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // Go as far left as possible from the current cursor.
+        self.push_left_chain();
+
+        // Pop the top of the stack = next in-order node.
+        if self.sp == 0 {
+            return None;
+        }
+        self.sp -= 1;
+        let i = self.stack[self.sp];
+        let n = &self.arena[i];
+
+        // After visiting `i`, the next subtree to process is its right child.
+        self.cursor = n.r();
+
+        Some((n.key, n.val))
+    }
+}
+
+// Optional ergonomic IntoIterator for &AvlMap
+impl<'a, K: Ord + Copy, V: Copy, T: Copy> IntoIterator for &'a AvlTreeMapView<'a, K, V, T> {
+    type Item = (K, V);
+    type IntoIter = AvlIter<'a, K, V, T>;
+    fn into_iter(self) -> Self::IntoIter {
+        self.iter()
+    }
+}
```

Comment:
I don't see any usages of this in `fsm.rs` or in the tests.

Maybe I missed it, but in general it would be good to delete anything which isn't used by the existing code, just so that we don't assume it has test coverage / existing-usage in future and start using it without either of those things.